### PR TITLE
mame: 0.226 -> 0.237

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

Our MAME derivation got kinda old, so bumping it.

As a bonus, one less program depending on Python 2 to build.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/misc/emulators/mame/default.nix [hunk 1]
```diff
@@ -34,17 +47,16 @@ in mkDerivation {
     "TOOLS=1"
     "USE_LIBSDL=1"
   ]
-  ++ lib.optionals stdenv.cc.isClang [ "CC=clang" "CXX=clang++" ]
-  ;
+  ++ lib.optionals stdenv.cc.isClang [ "CC=clang" "CXX=clang++" ];
```
Comment (r-burns) [inline pkgs/misc/emulators/mame/default.nix:50]: how about unconditionally providing `CC=${stdenv.cc.targetPrefix}cc CXX=${stdenv.cc.targetPrefix}c++`

Comment (thiagokokada) [reply pkgs/misc/emulators/mame/default.nix:50]: I don't think this would work:

```
nix-repl> stdenv.cc.targetPrefix
""
```

Comment (r-burns) [reply pkgs/misc/emulators/mame/default.nix:50]: Yeah that's right for the native case. Does the makefile not like `CC=cc`?

**Comment by thiagokokada** [comment]
Huh... This failed with a strange error: http://ix.io/3FDt

File: pkgs/misc/emulators/mame/default.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/misc/emulators/mame/default.nix:50]: Got this on macOS:

```
nix-repl> stdenv.cc.targetPrefix
""
```

Or unless this is something that is only changed on `stdenv.mkDerivation` :thinking: ?

**Comment by r-burns** [comment]
I see:

```
../../../../../src/osd/modules/sound/pulse_sound.cpp:24:10: fatal error: pulse/pulseaudio.h: No such file or directory
   24 | #include <pulse/pulseaudio.h>
      |          ^~~~~~~~~~~~~~~~~~~~
```

Did they add a new dependency on pulseaudio?

**Comment by thiagokokada** [comment]
> I see:
> 
> ```
> ../../../../../src/osd/modules/sound/pulse_sound.cpp:24:10: fatal error: pulse/pulseaudio.h: No such file or directory
>    24 | #include <pulse/pulseaudio.h>
>       |          ^~~~~~~~~~~~~~~~~~~~
> ```
> 
> Did they add a new dependency on pulseaudio?

Good catch. Let me try to add it to Linux compilation.

File: pkgs/misc/emulators/mame/default.nix (hunk 1 above)
Comment (r-burns) [reply pkgs/misc/emulators/mame/default.nix:50]: Yeah that's correct. Whenever not cross-compiling, targetPrefix is an empty string. The point of providing it to gracefully handle cross-compilation scenarios; e.g. when cross-compiling to i686-linux, targetPrefix will be "i686-unknown-linux-gnu-".

TLDR ${stdenv.cc.targetPrefix}cc is the most generic way to provide the name of the C compiler, and should work in all cases

Comment (thiagokokada) [reply pkgs/misc/emulators/mame/default.nix]: Ok, replaced the old CC/CXX flags with those ones.

Let me test if this will work on macOS...

**Comment by r-burns** [comment]
re: pulseaudio, it looks like support was added in https://github.com/mamedev/mame/commit/9c8b72a5febb9c1e1a038a0105c5e35f4bc49a0d, but it was made optional again in https://github.com/mamedev/mame/commit/52615c0aca4aa80cdb6c7f54645100766c6043fd. We build most applications with pulse support in nixpkgs anyway so it makes sense to add it as a new dependency IMO :+1:

File: pkgs/misc/emulators/mame/default.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/misc/emulators/mame/default.nix]: Got the following error on macOS:

http://ix.io/3FDL

Strange, I thought our default Clang was 7.0 :thinking: .

**Comment by thiagokokada** [comment]
> re: pulseaudio, it looks like support was added in [mamedev/mame@9c8b72a](https://github.com/mamedev/mame/commit/9c8b72a5febb9c1e1a038a0105c5e35f4bc49a0d), but it was made optional again in [mamedev/mame@52615c0](https://github.com/mamedev/mame/commit/52615c0aca4aa80cdb6c7f54645100766c6043fd). We build most applications with pulse support in nixpkgs anyway so it makes sense to add it as a new dependency IMO +1

There is this page showing all dependencies necessary to build MAME: https://docs.mamedev.org/initialsetup/compilingmame.html

I just added all of them that were missing.

File: pkgs/misc/emulators/mame/default.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/misc/emulators/mame/default.nix]: It seems to be using Clang 5.0.

Can I force Clang 6.0 or 7.0 without changing to `clangStdenv` :thinking: ? Only for macOS obviously.

Comment (r-burns) [reply pkgs/misc/emulators/mame/default.nix]: Whoa that's weird, I can't find where that's override happens. The clang5 override has to be somewhere, our stdenv is clang7 normally on x86_64-darwin

Comment (r-burns) [reply pkgs/misc/emulators/mame/default.nix]: Oh never mind, I think I see the issue. qt5's internal mkDerivation uses clang5 for some reason. Should be as simple as this:

```diff
--- a/pkgs/misc/emulators/mame/default.nix
+++ b/pkgs/misc/emulators/mame/default.nix
@@ -33,7 +33,7 @@ let
 
   dest = "$out/opt/mame";
 in
-mkDerivation rec {
+stdenv.mkDerivation rec {
   pname = "mame";
   version = "0.237";
 
```

**Comment by r-burns** [comment]
@ofborg build mame

**Comment by thiagokokada** [comment]
macOS is still failing: http://ix.io/3FE2

**Comment by r-burns** [comment]
Sure, we could try that. I see:

```diff
In file included from ../../../../../src/emu/http.cpp:18:
In file included from ../../../../../src/lib/util/server_ws_impl.hpp:9:
In file included from ../../../../../src/osd/asio.h:34:
In file included from ../../../../../3rdparty/asio/include/asio.hpp:21:
In file included from ../../../../../3rdparty/asio/include/asio/basic_datagram_socket.hpp:20:
In file included from ../../../../../3rdparty/asio/include/asio/basic_socket.hpp:40:
In file included from ../../../../../3rdparty/asio/include/asio/detail/reactive_socket_service.hpp:22:
In file included from ../../../../../3rdparty/asio/include/asio/buffer.hpp:27:
In file included from ../../../../../3rdparty/asio/include/asio/detail/string_view.hpp:23:
/nix/store/78a4mwr6n6jfnjs6k26jbm56xyaarihj-libcxx-7.1.0-dev/include/c++/v1/experimental/string_view:11:2: error: "<experimental/string_view> has been removed. Use <string_view> instead."
#error "<experimental/string_view> has been removed. Use <string_view> instead."
 ^
In file included from ../../../../../src/emu/http.cpp:18:
In file included from ../../../../../src/lib/util/server_ws_impl.hpp:9:
In file included from ../../../../../src/osd/asio.h:34:
In file included from ../../../../../3rdparty/asio/include/asio.hpp:21:
In file included from ../../../../../3rdparty/asio/include/asio/basic_datagram_socket.hpp:20:
In file included from ../../../../../3rdparty/asio/include/asio/basic_socket.hpp:40:
In file included from ../../../../../3rdparty/asio/include/asio/detail/reactive_socket_service.hpp:22:
In file included from ../../../../../3rdparty/asio/include/asio/buffer.hpp:27:
../../../../../3rdparty/asio/include/asio/detail/string_view.hpp:31:12: error: no member named 'experimental' in namespace 'std'
using std::experimental::basic_string_view;
      ~~~~~^
../../../../../3rdparty/asio/include/asio/detail/string_view.hpp:32:12: error: no member named 'experimental' in namespace 'std'
using std::experimental::string_view;
      ~~~~~^
```

I've seen this before, it's an issue in boost asio which has been addressed in recent versions. Unfortunately as I recall the fix didn't backport cleanly, at least for the boost version I was looking at. So if clang 6 doesn't work, another option would be to try to unvendor this bundled boost, to use a more recent version that contains that fix.

**Comment by thiagokokada** [comment]
@ofborg build mame

**Comment by r-burns** [comment]
Error:

```
../../../../../src/lib/util/server_ws_impl.hpp:364:74: error: 'using element_type = class asio::basic_stream_socket<asio::ip::tcp>' {aka 'class asio::basic_stream_socket<asio::ip::tcp>'} has no member named 'get_io_service'
  364 |    auto timer = std::make_shared<asio::system_timer>(connection->socket->get_io_service());
      |                                                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
```

We're at asio v1.16, but get_io_service was removed in 1.13
https://github.com/boostorg/asio/commit/a72fbb0b867fd8fb37a233d893c3f5e063e72ea6

We have `asio_1_10` but it's broken on darwin >:(

**Comment by thiagokokada** [comment]
> Error:
> 
> ```
> ../../../../../src/lib/util/server_ws_impl.hpp:364:74: error: 'using element_type = class asio::basic_stream_socket<asio::ip::tcp>' {aka 'class asio::basic_stream_socket<asio::ip::tcp>'} has no member named 'get_io_service'
>   364 |    auto timer = std::make_shared<asio::system_timer>(connection->socket->get_io_service());
>       |                                                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
> ```
> 
> We're at asio v1.16, but get_io_service was removed in 1.13 [chriskohlhoff/asio@2c689ad](https://github.com/chriskohlhoff/asio/commit/2c689adc9036441530b98b55a62856a7f0fb7d34)
> 
> We have `asio_1_10` but it's broken on darwin >:(

MAME `master` upgrade asio to 1.20, so we can either:
- Use Clang 6.0 for now, and once 0.238 is releases we can try Clang 7.0+ again.
- Backport the upgrade patches from 1.20 and see if it works.
- Wait for the 0.238.

I think I will go with just using Clang 6.0 for now since it seems the release of MAME 0.238 is really close. We can remove the Clang 6.0 override after MAME 0.238 is released.

**Comment by thiagokokada** [comment]
@ofborg build mame

**Comment by thiagokokada** [comment]
@ofborg build mame

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 146831` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mame (mess ,sdlmame)</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Tested with 1941 rom, seems fine.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 146831` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mame</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
I can't test on macOS since it is failing with:

```
2021-11-21 00:29:20.482 mame[64434:1657403] ../../../../../3rdparty/bgfx/src/glcontext_nsgl.mm (209): BGFX 0x00000002: Failed to initialize pixel format.
Abort trap: 6
```

However, I highly suspect this issue is caused by my VM (there is no 3D acceleration on it) and it should work fine on real hardware.

So I think this PR is ready to review. CC @r-burns.

**Review by r-burns — APPROVED** [review]
Cool, ofborg is happy on linux

Result of `nixpkgs-review pr 146831` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mame</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Thanks for the help @r-burns, merging it them.

In case macOS is really broken I hope merging this will make people aware of it.
